File: internal/config/orgchart.go
package config

import (
	"sync"

	"github.com/Z-M-Huang/openhive/internal/domain"
)

// OrgChartService implements the domain.OrgChart interface.
// It maintains an in-memory cache of the agent/team hierarchy
// built from all config files.
type OrgChartService struct {
	teams      map[string]*domain.Team   // keyed by slug
	agents     map[string]*domain.Agent  // keyed by AID
	agentTeam  map[string]string         // AID -> team slug
	leadTeams  map[string][]string       // AID -> team slugs led
	mu         sync.RWMutex
}

// NewOrgChart creates a new OrgChart service.
func NewOrgChart() *OrgChartService {
	return &OrgChartService{
		teams:     make(map[string]*domain.Team),
		agents:    make(map[string]*domain.Agent),
		agentTeam: make(map[string]string),
		leadTeams: make(map[string][]string),
	}
}

// RebuildFromConfig rebuilds the OrgChart cache from master config and team configs.
func (o *OrgChartService) RebuildFromConfig(master *domain.MasterConfig, teams map[string]*domain.Team) error {
	o.mu.Lock()
	defer o.mu.Unlock()

	// Reset maps
	o.teams = make(map[string]*domain.Team)
	o.agents = make(map[string]*domain.Agent)
	o.agentTeam = make(map[string]string)
	o.leadTeams = make(map[string][]string)

	// Track all AIDs for uniqueness check
	allAIDs := make(map[string]string) // AID -> source (team slug or "master")

	// Add main assistant
	if master.Assistant.AID != "" {
		agent := &domain.Agent{
			AID:            master.Assistant.AID,
			Name:           master.Assistant.Name,
			Provider:       master.Assistant.Provider,
			ModelTier:      master.Assistant.ModelTier,
			MaxTurns:       master.Assistant.MaxTurns,
			TimeoutMinutes: master.Assistant.TimeoutMinutes,
		}
		o.agents[agent.AID] = agent
		allAIDs[agent.AID] = "master"
	}

	// Add top-level team lead agents from master config
	for i := range master.Agents {
		agent := &master.Agents[i]
		if _, exists := allAIDs[agent.AID]; exists {
			return &domain.ConflictError{
				Resource: "agent",
				Message:  "duplicate AID " + agent.AID + " in master config",
			}
		}
		o.agents[agent.AID] = agent
		allAIDs[agent.AID] = "master"
	}

	// Process teams
	for slug, team := range teams {
		team.Slug = slug
		o.teams[slug] = team

		// Register leader
		if team.LeaderAID != "" {
			o.leadTeams[team.LeaderAID] = append(o.leadTeams[team.LeaderAID], slug)
		}

		// Register team agents
		for i := range team.Agents {
			agent := &team.Agents[i]
			if existingSource, exists := allAIDs[agent.AID]; exists {
				return &domain.ConflictError{
					Resource: "agent",
					Message:  "duplicate AID " + agent.AID + " (in " + slug + " and " + existingSource + ")",
				}
			}
			o.agents[agent.AID] = agent
			o.agentTeam[agent.AID] = slug
			allAIDs[agent.AID] = slug
		}
	}

	// Validate: check for circular parents
	for slug, team := range o.teams {
		if err := o.detectCircularParent(slug, team.ParentSlug); err != nil {
			return err
		}
	}

	return nil
}

func (o *OrgChartService) detectCircularParent(startSlug, parentSlug string) error {
	visited := map[string]bool{startSlug: true}
	current := parentSlug
	for current != "" {
		if visited[current] {
			return &domain.ValidationError{
				Field:   "parent_slug",
				Message: "circular parent chain detected involving " + startSlug,
			}
		}
		visited[current] = true
		team, exists := o.teams[current]
		if !exists {
			break
		}
		current = team.ParentSlug
	}
	return nil
}

// GetOrgChart returns all teams in the hierarchy.
func (o *OrgChartService) GetOrgChart() map[string]*domain.Team {
	o.mu.RLock()
	defer o.mu.RUnlock()
	result := make(map[string]*domain.Team, len(o.teams))
	for k, v := range o.teams {
		result[k] = v
	}
	return result
}

// GetAgentByAID returns an agent by its AID.
func (o *OrgChartService) GetAgentByAID(aid string) (*domain.Agent, error) {
	o.mu.RLock()
	defer o.mu.RUnlock()
	agent, exists := o.agents[aid]
	if !exists {
		return nil, &domain.NotFoundError{Resource: "agent", ID: aid}
	}
	return agent, nil
}

// GetTeamBySlug returns a team by its slug.
func (o *OrgChartService) GetTeamBySlug(slug string) (*domain.Team, error) {
	o.mu.RLock()
	defer o.mu.RUnlock()
	team, exists := o.teams[slug]
	if !exists {
		return nil, &domain.NotFoundError{Resource: "team", ID: slug}
	}
	return team, nil
}

// GetTeamForAgent returns the team an agent belongs to.
func (o *OrgChartService) GetTeamForAgent(aid string) (*domain.Team, error) {
	o.mu.RLock()
	defer o.mu.RUnlock()
	slug, exists := o.agentTeam[aid]
	if !exists {
		return nil, &domain.NotFoundError{Resource: "team for agent", ID: aid}
	}
	team, exists := o.teams[slug]
	if !exists {
		return nil, &domain.NotFoundError{Resource: "team", ID: slug}
	}
	return team, nil
}

// GetLeadTeams returns the team slugs that an agent leads.
func (o *OrgChartService) GetLeadTeams(aid string) ([]string, error) {
	o.mu.RLock()
	defer o.mu.RUnlock()
	slugs, exists := o.leadTeams[aid]
	if !exists {
		return nil, nil
	}
	result := make([]string, len(slugs))
	copy(result, slugs)
	return result, nil
}

// GetSubordinates returns agents that report to the given agent.
func (o *OrgChartService) GetSubordinates(aid string) ([]domain.Agent, error) {
	o.mu.RLock()
	defer o.mu.RUnlock()

	teamSlugs := o.leadTeams[aid]
	var subordinates []domain.Agent
	for _, slug := range teamSlugs {
		team, exists := o.teams[slug]
		if !exists {
			continue
		}
		subordinates = append(subordinates, team.Agents...)
	}
	return subordinates, nil
}

// GetSupervisor returns the lead agent of the team this agent belongs to.
func (o *OrgChartService) GetSupervisor(aid string) (*domain.Agent, error) {
	o.mu.RLock()
	defer o.mu.RUnlock()

	slug, exists := o.agentTeam[aid]
	if !exists {
		return nil, &domain.NotFoundError{Resource: "supervisor for agent", ID: aid}
	}
	team, exists := o.teams[slug]
	if !exists {
		return nil, &domain.NotFoundError{Resource: "team", ID: slug}
	}
	leader, exists := o.agents[team.LeaderAID]
	if !exists {
		return nil, &domain.NotFoundError{Resource: "leader agent", ID: team.LeaderAID}
	}
	return leader, nil
}

File: backend/src/config/orgchart.ts
/**
 * OpenHive Backend - OrgChartService
 *
 * In-memory hierarchy cache built from config files. Tracks teams, agents,
 * team membership, lead assignments, and detects circular parent chains.
 *
 * All methods are synchronous — the cache is rebuilt atomically via
 * rebuildFromConfig() and all subsequent lookups are pure Map reads.
 */

import type { Team, Agent, MasterConfig } from '../domain/types.js';
import type { OrgChart } from '../domain/interfaces.js';
import { NotFoundError, ConflictError, ValidationError } from '../domain/errors.js';

// ---------------------------------------------------------------------------
// OrgChartService
// ---------------------------------------------------------------------------

/**
 * Implements the OrgChart interface.
 * Maintains an in-memory index of the agent/team hierarchy
 * built from all config files.
 */
export class OrgChartService implements OrgChart {
  /** teams keyed by slug */
  private teams: Map<string, Team> = new Map();

  /** agents keyed by AID */
  private agents: Map<string, Agent> = new Map();

  /** AID -> team slug (which team this agent is a member of) */
  private agentTeam: Map<string, string> = new Map();

  /** AID -> team slugs that this agent leads */
  private leadTeams: Map<string, string[]> = new Map();

  // -------------------------------------------------------------------------
  // rebuildFromConfig
  // -------------------------------------------------------------------------

  /**
   * Rebuilds the OrgChart cache from master config and team configs.
   *
   * Steps:
   *   1. Reset all internal maps.
   *   2. Register main assistant (if AID is set).
   *   3. Register top-level agents from master.agents (check for duplicate AIDs).
   *   4. Process each team's agents (check duplicates, register leadTeams).
   *   5. Validate for circular parent chains.
   *
   * Throws ConflictError on duplicate AID.
   * Throws ValidationError on circular parent chain.
   */
  rebuildFromConfig(master: MasterConfig, teams: Record<string, Team>): void {
    // Reset maps
    this.teams = new Map();
    this.agents = new Map();
    this.agentTeam = new Map();
    this.leadTeams = new Map();

    // Track all AIDs for uniqueness check: AID -> source (team slug or "master")
    const allAIDs = new Map<string, string>();

    // Register synthetic 'main' team for the main container.
    // The main assistant and top-level team leaders run in the main container.
    // This team has no config file — it's derived from master config.
    this.teams.set('main', {
      tid: '',
      slug: 'main',
      leader_aid: master.assistant.aid,
    });

    // Register main assistant
    if (master.assistant.aid !== '') {
      const assistant: Agent = {
        aid: master.assistant.aid,
        name: master.assistant.name,
        provider: master.assistant.provider,
        model_tier: master.assistant.model_tier,
        max_turns: master.assistant.max_turns,
        timeout_minutes: master.assistant.timeout_minutes,
      };
      this.agents.set(assistant.aid, assistant);
      this.agentTeam.set(assistant.aid, 'main');
      allAIDs.set(assistant.aid, 'master');
    }

    // Register top-level team lead agents from master config.
    // These agents run in the main container, so map them to 'main' in
    // agentTeam so getTeamForAgent can route tasks to them.
    if (master.agents !== undefined) {
      for (const agent of master.agents) {
        if (allAIDs.has(agent.aid)) {
          throw new ConflictError('agent', `duplicate AID ${agent.aid} in master config`);
        }
        this.agents.set(agent.aid, agent);
        this.agentTeam.set(agent.aid, 'main');
        allAIDs.set(agent.aid, 'master');
      }
    }

    // Process teams
    for (const [slug, team] of Object.entries(teams)) {
      // Ensure slug is set on the team
      const teamWithSlug: Team = { ...team, slug };
      this.teams.set(slug, teamWithSlug);

      // Register leader in leadTeams
      if (teamWithSlug.leader_aid !== '') {
        const existing = this.leadTeams.get(teamWithSlug.leader_aid);
        if (existing !== undefined) {
          existing.push(slug);
        } else {
          this.leadTeams.set(teamWithSlug.leader_aid, [slug]);
        }
      }

      // Register team agents
      if (teamWithSlug.agents !== undefined) {
        for (const agent of teamWithSlug.agents) {
          const existingSource = allAIDs.get(agent.aid);
          if (existingSource !== undefined) {
            throw new ConflictError(
              'agent',
              `duplicate AID ${agent.aid} (in ${slug} and ${existingSource})`,
            );
          }
          this.agents.set(agent.aid, agent);
          this.agentTeam.set(agent.aid, slug);
          allAIDs.set(agent.aid, slug);
        }
      }
    }

    // Validate: check for circular parents
    for (const [slug, team] of this.teams.entries()) {
      this.detectCircularParent(slug, team.parent_slug ?? '');
    }
  }

  // -------------------------------------------------------------------------
  // detectCircularParent
  // -------------------------------------------------------------------------

  /**
   * Traverses the parent chain from startSlug to detect cycles.
   *
   * Throws ValidationError if a cycle is detected.
   */
  private detectCircularParent(startSlug: string, parentSlug: string): void {
    const visited = new Set<string>([startSlug]);
    let current = parentSlug;
    while (current !== '') {
      if (visited.has(current)) {
        throw new ValidationError(
          'parent_slug',
          `circular parent chain detected involving ${startSlug}`,
        );
      }
      visited.add(current);
      const team = this.teams.get(current);
      if (team === undefined) {
        break;
      }
      current = team.parent_slug ?? '';
    }
  }

  // -------------------------------------------------------------------------
  // getOrgChart
  // -------------------------------------------------------------------------

  /**
   * Returns all teams in the hierarchy as a plain object keyed by slug.
   *
   * Returns a shallow copy of the internal map.
   */
  getOrgChart(): Record<string, Team> {
    const result: Record<string, Team> = {};
    for (const [k, v] of this.teams.entries()) {
      // Exclude the synthetic 'main' team — it's an internal routing entry.
      if (k === 'main') continue;
      result[k] = v;
    }
    return result;
  }

  // -------------------------------------------------------------------------
  // getAgentByAID
  // -------------------------------------------------------------------------

  /**
   * Returns an agent by its AID.
   *
   * Throws NotFoundError if the agent does not exist.
   */
  getAgentByAID(aid: string): Agent {
    const agent = this.agents.get(aid);
    if (agent === undefined) {
      throw new NotFoundError('agent', aid);
    }
    return agent;
  }

  // -------------------------------------------------------------------------
  // getTeamBySlug
  // -------------------------------------------------------------------------

  /**
   * Returns a team by its slug.
   *
   * Throws NotFoundError if the team does not exist.
   */
  getTeamBySlug(slug: string): Team {
    const team = this.teams.get(slug);
    if (team === undefined) {
      throw new NotFoundError('team', slug);
    }
    return team;
  }

  // -------------------------------------------------------------------------
  // getTeamForAgent
  // -------------------------------------------------------------------------

  /**
   * Returns the team an agent belongs to.
   *
   * Throws NotFoundError if the agent has no team mapping or the team is missing.
   */
  getTeamForAgent(aid: string): Team {
    const slug = this.agentTeam.get(aid);
    if (slug === undefined) {
      throw new NotFoundError('team for agent', aid);
    }
    const team = this.teams.get(slug);
    if (team === undefined) {
      throw new NotFoundError('team', slug);
    }
    return team;
  }

  // -------------------------------------------------------------------------
  // getLeadTeams
  // -------------------------------------------------------------------------

  /**
   * Returns the team slugs that an agent leads.
   *
   * Returns an empty array if the agent does not lead any teams.
   */
  getLeadTeams(aid: string): string[] {
    const slugs = this.leadTeams.get(aid);
    if (slugs === undefined) {
      return [];
    }
    return [...slugs];
  }

  // -------------------------------------------------------------------------
  // getSubordinates
  // -------------------------------------------------------------------------

  /**
   * Returns agents that report to the given agent (i.e. members of all teams
   * the given agent leads).
   *
   * Returns an empty array if the agent leads no teams.
   */
  getSubordinates(aid: string): Agent[] {
    const teamSlugs = this.leadTeams.get(aid);
    if (teamSlugs === undefined) {
      return [];
    }
    const subordinates: Agent[] = [];
    for (const slug of teamSlugs) {
      const team = this.teams.get(slug);
      if (team === undefined || team.agents === undefined) {
        continue;
      }
      subordinates.push(...team.agents);
    }
    return subordinates;
  }

  // -------------------------------------------------------------------------
  // getSupervisor
  // -------------------------------------------------------------------------

  /**
   * Returns the lead agent of the team this agent belongs to.
   *
   * Returns null if:
   *   - the agent has no team membership (top-level or master agent)
   *   - the team's leader_aid does not resolve to a known agent
   */
  getSupervisor(aid: string): Agent | null {
    const slug = this.agentTeam.get(aid);
    if (slug === undefined) {
      return null;
    }
    const team = this.teams.get(slug);
    if (team === undefined) {
      return null;
    }
    // An agent cannot be its own supervisor.
    if (team.leader_aid === aid) {
      return null;
    }
    const leader = this.agents.get(team.leader_aid);
    if (leader === undefined) {
      return null;
    }
    return leader;
  }
}

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

/**
 * Creates a new OrgChartService instance.
 */
export function newOrgChart(): OrgChartService {
  return new OrgChartService();
}
